Wikipedia/Code/WebViewController.m

```diff
@@ -249,6 +262,142 @@ - (void)handleArticleStateScriptMessage:(NSString*)messageString {
     }
 }
 
+- (void)handleFindInPageMatchesFoundMessage:(NSArray*)messageArray {
+    self.findInPageMatches = messageArray;
+    self.findInPageResultCursorIndex = -1;
+}
+
+#pragma mark - Find-in-page
+
+- (FindInPageKeyboardBar *)findInPageKeyboardBar {
+    return self.view.inputAccessoryView;
+}
+
+- (UITextField *)findInPageTextField {
+    return [self findInPageKeyboardBar].textField;
+}
+
+- (BOOL)isFindInPageVisible {
+    return [[self findInPageTextField] isFirstResponder];
+}
+
+- (void)showFindInPage {
+    [self referencesHide];
+    [self.view becomeFirstResponder];
+    [[self findInPageTextField] becomeFirstResponder];
+}
+
+- (void)hideFindInPage {
+    [self clearFindInPage];
+    [[self findInPageTextField] resignFirstResponder];
+    [self.view resignFirstResponder];
+    [self.webView evaluateJavaScript:@"window.wmf.findInPage.removeSearchTermHighlights()"
+                   completionHandler:nil];
+}
+
+- (void)clearFindInPage {
+    self.findInPageMatches = @[];
+    self.findInPageResultCursorIndex = -1;
+    [[self findInPageTextField] setText:@""];
+}
+
+- (void)minimizeFindInPage {
+    if (!self.disableMinimizeFindInPage) {
+        [[self findInPageTextField] resignFirstResponder];
+    }
+}
+
+- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
+    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
+    self.disableMinimizeFindInPage = YES;
+    [coordinator animateAlongsideTransition:nil completion:^(id<UIViewControllerTransitionCoordinatorContext> context) {
+        self.disableMinimizeFindInPage = NO;
+    }];
+}
+
+- (void)setFindInPageMatches:(NSArray *)findInPageMatches {
+    _findInPageMatches = findInPageMatches;
+    [self findInPageKeyboardBar].numberOfMatches = [findInPageMatches count];
+}
+
+#pragma FindInPageBar cursor
+
+- (void)setFindInPageResultCursorIndex:(NSInteger)findInPageResultCursorIndex {
+    _findInPageResultCursorIndex = findInPageResultCursorIndex;
+    [self findInPageKeyboardBar].currentCursorIndex = findInPageResultCursorIndex;
+}
+
+- (void)advanceFindInPageResultCursorIndexReversed:(BOOL)reversed {
```

Comment:
A little nit picky, but having a method start with the word "advance" and end in "reversed" sounds weird/clunky and hurts readability.

I usually use a pair of methods to make the intent more clear like:
`selectNextFindInPageResult` and `selectPreviousFindInPageResult` 

These would usually call through to a shared method… 
`selectFindInPageResultAtIndex:` 
with something like: 
`[self selectFindInPageResultAtIndex:self.findInPageResultCursorIndex+1]`